sdk/schemaregistry/azure-data-schemaregistry-apacheavro/src/main/java/com/azure/data/schemaregistry/apacheavro/SchemaRegistryApacheAvroEncoder.java

```diff
@@ -0,0 +1,335 @@
+        }
+
+        final String schemaFullName = schema.getFullName();
+        final String schemaString = schema.toString();
+
+        return maybeRegisterSchema(serializerOptions.getSchemaGroup(), schemaFullName, schemaString)
+            .handle((schemaId, sink) -> {
+                try {
+                    final byte[] encoded = avroSerializer.encode(object);
+                    final T serializedMessage = messageFactoryToUse.apply(BinaryData.fromBytes(encoded));
+
+                    serializedMessage.setContentType(AVRO_MIME_TYPE + "+" + schemaId);
+
+                    sink.next(serializedMessage);
+                    sink.complete();
+                } catch (Exception e) {
+                    sink.error(new RuntimeException(String.format(
+                        "Error encountered serializing object: %s with schemaId '%s'.", object, schemaId)));
+                }
+            });
+    }
+
+    /**
+     * Decodes a message into its object.
+     *
+     * @param message Object to encode.
+     * @param typeReference Message to encode to.
+     * @param <T> Concrete type of {@link MessageWithMetadata}.
+     *
+     * @return The message encoded.
+     */
+    public <T> T decodeMessageData(MessageWithMetadata message, TypeReference<T> typeReference) {
+        return decodeMessageDataAsync(message, typeReference).block(TIMEOUT);
+    }
+
+    /**
+     * Decodes a message into its object.
+     *
+     * @param message Object to encode.
+     * @param typeReference Message to encode to.
+     * @param <T> Concrete type of {@link MessageWithMetadata}.
+     *
+     * @return The message encoded.
+     * @throws NullPointerException if {@code message} or {@code typeReference} is null.
+     */
+    public <T> Mono<T> decodeMessageDataAsync(MessageWithMetadata message, TypeReference<T> typeReference) {
+        if (message == null) {
+            return monoError(logger, new NullPointerException("'message' cannot be null."));
+        } else if (typeReference == null) {
+            return monoError(logger, new NullPointerException("'typeReference' cannot be null."));
+        }
+
+        final BinaryData body = message.getBodyAsBinaryData();
+
+        if (Objects.isNull(body)) {
+            logger.warning("Message provided does not have a BinaryBody, returning empty response.");
+            return Mono.empty();
+        }
+
+        final ByteBuffer contents = body.toByteBuffer();
+
+        if (contents.remaining() == 0) {
+            logger.warning("Message provided has an empty BinaryBody, returning empty response.");
+            return Mono.empty();
+        }
+
+        final String schemaId;
+
+        // Temporary back-compat for the first beta while we phase this out. In the future, it will return an error.
+        // Check if the first 4 bytes of the payload have the format.
+        final byte[] recordFormatIndicator = new byte[RECORD_FORMAT_INDICATOR_SIZE];
+        contents.mark();
```

Comment:
Should we consider removing this eagerly? The user is already aware of breaking changes due to the type name (serializer -> encoder) and would expect to have breaking changes with the data too. Removing this back compat later can lead to users having to deal with breaking changes in this release and in the future release when we remove this support.